dot/core/service_test.go

```diff
@@ -586,3 +656,699 @@ func Test_Service_handleBlocksAsync(t *testing.T) {
+		expChanges := make(QueryKeyValueChanges)
+		expChanges["0x01"] = common.BytesToHex([]byte{1, 2, 3})
+		s := &Service{
+			storageState: mockStorageState,
+		}
+		execTest(t, s, common.Hash{}, []string{"0x01"}, expChanges, nil)
+	})
+}
+
+func TestService_QueryStorage(t *testing.T) {
+	t.Parallel()
+	execTest := func(t *testing.T, s *Service, from common.Hash, to common.Hash,
+		keys []string, exp map[common.Hash]QueryKeyValueChanges, expErr error) {
+		res, err := s.QueryStorage(from, to, keys...)
+		assert.ErrorIs(t, err, expErr)
+		if expErr != nil {
+			assert.EqualError(t, err, expErr.Error())
+		}
+		assert.Equal(t, exp, res)
+	}
+
+	t.Run("subchain error", func(t *testing.T) {
+		t.Parallel()
+		ctrl := gomock.NewController(t)
+		mockBlockState := NewMockBlockState(ctrl)
+		mockBlockState.EXPECT().BestBlockHash().Return(common.Hash{2})
+		mockBlockState.EXPECT().SubChain(common.Hash{1}, common.Hash{2}).Return(nil, errDummyErr)
+		s := &Service{
+			blockState: mockBlockState,
+		}
+		execTest(t, s, common.Hash{1}, common.Hash{}, nil, nil, errDummyErr)
+	})
+
+	t.Run("happy path", func(t *testing.T) {
+		t.Parallel()
+		ctrl := gomock.NewController(t)
+		mockBlockState := NewMockBlockState(ctrl)
+		mockBlockState.EXPECT().BestBlockHash().Return(common.Hash{2})
+		mockBlockState.EXPECT().SubChain(common.Hash{1}, common.Hash{2}).Return([]common.Hash{{0x01}}, nil)
+		mockStorageState := NewMockStorageState(ctrl)
+		mockStorageState.EXPECT().GetStateRootFromBlock(&common.Hash{0x01}).Return(&common.Hash{}, nil)
+		mockStorageState.EXPECT().GetStorage(&common.Hash{}, common.MustHexToBytes("0x01")).
+			Return([]byte{1, 2, 3}, nil)
+		expChanges := make(QueryKeyValueChanges)
+		expChanges["0x01"] = common.BytesToHex([]byte{1, 2, 3})
+		expQueries := make(map[common.Hash]QueryKeyValueChanges)
+		expQueries[common.Hash{0x01}] = expChanges
+		s := &Service{
+			blockState:   mockBlockState,
+			storageState: mockStorageState,
+		}
+		execTest(t, s, common.Hash{1}, common.Hash{}, []string{"0x01"}, expQueries, nil)
+	})
+}
+
+func TestService_GetReadProofAt(t *testing.T) {
+	t.Parallel()
+	execTest := func(t *testing.T, s *Service, block common.Hash, keys [][]byte,
+		expHash common.Hash, expProofForKeys [][]byte, expErr error) {
+		resHash, resProofForKeys, err := s.GetReadProofAt(block, keys)
+		assert.ErrorIs(t, err, expErr)
+		if expErr != nil {
+			assert.EqualError(t, err, expErr.Error())
+		}
+		assert.Equal(t, expHash, resHash)
+		assert.Equal(t, expProofForKeys, resProofForKeys)
+	}
+
+	t.Run("get block state root error", func(t *testing.T) {
+		t.Parallel()
+		ctrl := gomock.NewController(t)
+		mockBlockState := NewMockBlockState(ctrl)
+		mockBlockState.EXPECT().BestBlockHash().Return(common.Hash{2})
+		mockBlockState.EXPECT().GetBlockStateRoot(common.Hash{2}).Return(common.Hash{}, errDummyErr)
+		s := &Service{
+			blockState: mockBlockState,
+		}
+		execTest(t, s, common.Hash{}, nil, common.Hash{}, nil, errDummyErr)
```

Comment:
nit just a suggestion, I started doing this not too long ago and I think it's nice readability-wise; variable assign each argument so it's clear i.e. without an IDE what's what.

```suggestion

		blockHash := common.Hash{}
		keys := ([][]byte)(nil)
		expectedHash := common.Hash{}
		expectedProofForKeys := ([][]byte)(nil)
		expectedErr := errDummyErr
		execTest(t, s, blockHash, keys, expectedHash, expectedProofForKeys, expectedErr)
```

Also `s` for service is sus, call it `service`.